### PR TITLE
Remove most allocations from rrtmgp interface

### PR DESCRIPTION
**Pull request by jgfouca**
Switched the rrtmgp interface in eamxx to make maximal use of the pool allocator I wrote for rrtmgp standalone.

I just got a very nice performance result for this branch on pm-gpu:

```
PASS SMS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.pm-gpu_gnugpu RUN time=157   # YAKL
PASS SMS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.pm-gpu_gnugpu RUN time=163   # Kokkos (before this PR)
PASS SMS_Ln362.ne30pg2_ne30pg2.F2010-SCREAMv1.pm-gpu_gnugpu RUN time=143   # Kokkos (with this PR)
```

### CODE REVIEW

File: components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp [hunk 1]
```diff
@@ -565,6 +565,44 @@ static void rrtmgp_sw(
...
+  const int size1 = nday;
+  const int size2 = nday*nlay; // 4
+  const int size3 = nday*(nlay+1); // 5
+  const int size4 = ncol*nlay;
+  const int size5 = nbnd*nday; //2
+  const int size6 = nday*ngpt;
+  const int size7 = nday*(nlay+1)*nbnd; // 3
+  const int size8 = ncol*nlay*(k_dist.get_ngas()+1);
+
+  RealT* data = pool_t::template alloc_raw<RealT>(size1 + size2*4 + size3*5 + size4 + size5*2 + size6 + size7*3 + size8), *dcurr = data;
```
Comment (bartgol) [inline components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:578]: From what I can tell, rrtmgp_sw is called by rrtmgp_main, so we'll still be allocating the pool at runtime. Is that correct? Can we maybe allocate the pool once at init, pass it around at runtime, and deallocate it inside finalize?

File: components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp [hunk 2]
```diff
@@ -819,6 +840,24 @@ static void rrtmgp_lw(
...
+  int constexpr max_gauss_pts = 4;
+
+  const int size1 = ncol;
+  const int size2 = nbnd*ncol;
+  const int size3 = max_gauss_pts*max_gauss_pts;
+  const int size4 = ncol*nlay;
+  const int size5 = ncol*(nlay+1);
+  const int size6 = ncol*nlay*(k_dist.get_ngas()+1);
+
+  RealT* data = pool_t::template alloc_raw<RealT>(size1 + size2 + size3*2 + size4 + size5 + size6), *dcurr = data;
```
Comment (bartgol) [inline components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:852]: Same here.

File: components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp [hunk 3]
```diff
@@ -882,32 +919,31 @@ static void rrtmgp_lw(
...
+  hview_t<RealT**> gauss_Ds_host (&gauss_Ds_host_raw[0][0], max_gauss_pts, max_gauss_pts);
+
+  RealT gauss_wts_host_raw[max_gauss_pts][max_gauss_pts] = {
+    {0.5, 0.3180413817, 0.2009319137, 0.1355069134},
+    {0., 0.1819586183, 0.2292411064, 0.2034645680},
+    {0., 0., 0.0698269799, 0.1298475476},
+    {0., 0., 0., 0.0311809710}
+  };
+
+  hview_t<RealT**> gauss_wts_host(&gauss_wts_host_raw[0][0],max_gauss_pts,max_gauss_pts);
```
Comment (bartgol) [inline components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:937]: Are we copying the same gauss quadrature info to device over and over at every time step? Can we do this at init as well, and then pass around the pre-filled views at runtime? It seems silly to set them up every time...

**Review by bartgol — COMMENTED** [review]
You decide how much to do in a single PR Jim, but I think there could be other unnecessary alloc's we could remove, as well as a pointless setup of gauss quadrature data at every time step.

File: components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp (hunk 3 above)
Comment (jgfouca) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:937]: We could do this at init but I don't think this is very expensive at all.

Comment (bartgol) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:937]: I agree that it's not very expensive, but so is a single device allocation (in the grand scheme of things). It's just something we can remove, and these small opt all pile up. But yeah, no need to do it here.

Comment (jgfouca) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:937]: @bartgol , i don't think there's any allocation happening here. The C arrays will be on the stack and the views just point to that memory.

Comment (bartgol) [reply components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:937]: There's no alloc, correct, but there is are two deep_copy, which involve a small kernel launch. Nothing big, but also pointless.

**Pull request merged**